### PR TITLE
update apply_to_collections to support dataclass inputs

### LINKED ISSUE #11759: `apply_to_collections` fails on dataclass inputs
## 🐛 Bug

If `apply_to_collections` is called on two dataclasses, the passed `function` will be called on only one of the input dataclasses. 

Dataclass inputs for `apply_to_collection` are handled by a conditional:

https://github.com/PyTorchLightning/pytorch-lightning/blob/8ddf9f996d4a98e7be5fc591b7734b6f78313771/pytorch_lightning/utilities/apply_func.py#L128

`apply_to_collections` does not appear to have such a conditional.

### To Reproduce
Code

```python
import torch
from torch import Tensor
from dataclasses import dataclass
from pytorch_lightning.utilities.apply_func import apply_to_collections

@dataclass
class Dataclass:
    x: Tensor = torch.rand(10)

dc1 = Dataclass()
dc2 = Dataclass()
apply_to_collections(dc1, dc2, Tensor, torch.allclose)
```

Fails with

```
Traceback (most recent call last):
  File "/home/tidal/Documents/dl-template/combustion/dc.py", line 18, in <module>
    apply_to_collections(dc1, dc2, Tensor, torch.allclose)
  File "/home/tidal/Documents/dl-template/combustion/venv/lib/python3.9/site-packages/pytorch_lightning/utilities/apply_func.py", line 215, in apply_to_collections
    return apply_to_collection(data1, dtype, function, *args, wrong_dtype=wrong_dtype, **kwargs)
  File "/home/tidal/Documents/dl-template/combustion/venv/lib/python3.9/site-packages/pytorch_lightning/utilities/apply_func.py", line 138, in apply_to_collection
    v = apply_to_collection(
  File "/home/tidal/Documents/dl-template/combustion/venv/lib/python3.9/site-packages/pytorch_lightning/utilities/apply_func.py", line 95, in apply_to_collection
    return function(data, *args, **kwargs)
TypeError: allclose() missing 1 required positional arguments: "other"
```

### Expected behavior

No exception should be raised, or it should be indicated that `apply_to_collections` does not support dataclasses.

### Environment

```
* CUDA:
        - GPU:
                - NVIDIA GeForce RTX 3090
                - NVIDIA GeForce RTX 3090
        - available:         True
        - version:           11.3
* Packages:
        - numpy:             1.22.1
        - pyTorch_debug:     False
        - pyTorch_version:   1.10.2+cu113
        - pytorch-lightning: 1.5.9
        - tqdm:              4.62.3
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:
        - python:            3.9.9
        - version:           #1 SMP PREEMPT Thu, 27 Jan 2022 18:27:00 +0000
```

cc @borda

### PR DESCRIPTION
**Pull request by armanal**
## What does this PR do?

Fixes #11759

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by armanal** [comment]
@awaelchli some tests are failing here. can you please help?

File: CHANGELOG.md [hunk 1]
```diff
@@ -590,6 +590,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Configure native Deepspeed schedulers with interval='step' ([#11788](https://github.com/PyTorchLightning/pytorch-lightning/pull/11788))
 
 
+- Fixed `apply_to_collections` to support inputs of type `dataclass` ([#11889](https://github.com/PyTorchLightning/pytorch-lightning/pull/11889))
```
Comment (awaelchli) [inline CHANGELOG.md:593]:
```suggestion
- Added support for dataclasses in `apply_to_collections` ([#11889](https://github.com/PyTorchLightning/pytorch-lightning/pull/11889))
```

Let's move this to the "Added" section?

File: pytorch_lightning/utilities/apply_func.py [hunk 2]
```diff
@@ -222,6 +222,47 @@ def apply_to_collections(
...
+                    **kwargs,
+                )
+            if not field_init1 or not field_init2 or v is None:  # retain old value
+                return apply_to_collection(data1, dtype, function, *args, wrong_dtype=wrong_dtype, **kwargs)
+            try:
+                setattr(result, field_name, v)
+            except dataclasses.FrozenInstanceError as e:
+                raise MisconfigurationException(
+                    "A frozen dataclass was passed to `apply_to_collection` but this is not allowed."
+                    " HINT: is your batch a frozen dataclass?"
```
Comment (awaelchli) [inline pytorch_lightning/utilities/apply_func.py:262]:
```suggestion
                    "A frozen dataclass was passed to `apply_to_collections` but this is not allowed."
                    " HINT: is your batch a frozen dataclass?"
```

File: tests/utilities/test_apply_func.py [hunk 3]
```diff
@@ -303,6 +303,60 @@ def fn(a, b):
...
+
+        def __eq__(self, o: object) -> bool:
+            if not isinstance(o, ModelExample):
+                return NotImplemented
+            else:
+                return (
+                    self.example_ids == o.example_ids
+                    and self.feature == o.feature
+                    and torch.equal(self.label, o.label)
+                    and self.some_constant == o.some_constant
```
Comment (awaelchli) [inline tests/utilities/test_apply_func.py:336]: these definitions are now duplicated. 
a small improvement could be to pull all definitions out of the `test_recursive_application_to_collection` test and provide them at module level to be used by all tests.
Do you think we should do this?

**Review by awaelchli — COMMENTED** [review]
Great! thanks for working on this

File: pytorch_lightning/utilities/apply_func.py [hunk 4]
```diff
@@ -222,6 +222,47 @@ def apply_to_collections(
         ]
         return elem_type(*out) if is_namedtuple else elem_type(out)
 
+    if _is_dataclass_instance(data1):
+        assert isinstance(data2, type(data1)), "Dataclasses are of different types"
```
Comment (tchaton) [inline pytorch_lightning/utilities/apply_func.py:226]: Should we assert the dataclasses have the same fields instead ?

File: tests/utilities/test_apply_func.py (hunk 3 above)
Comment (armanal) [reply tests/utilities/test_apply_func.py:336]: of course, that would be better. I'll move them.

File: pytorch_lightning/utilities/apply_func.py (hunk 4 above)
Comment (armanal) [reply pytorch_lightning/utilities/apply_func.py:226]: sure, I can do that instead. 
but I wonder, is it an expected situation where the user has defined two different types with the same fields and attempts to merge them together?
and, what should the type of reduced result be?

File: pytorch_lightning/utilities/apply_func.py [hunk 5]
```diff
@@ -222,6 +222,53 @@ def apply_to_collections(
         ]
         return elem_type(*out) if is_namedtuple else elem_type(out)
 
+    if _is_dataclass_instance(data1):
+        assert _is_dataclass_instance(data2), (
+            "Expected inputs to be dataclasses of the same type or to have identical fields"
+            f" but got input 1 of type {type(data1)} and input 2 of type {type(data2)}"
+        )
```
Comment (awaelchli) [inline pytorch_lightning/utilities/apply_func.py:229]: assertions are not meant for user input validation, let's raise an error instead :)

```suggestion
        if not _is_dataclass_instance(data2):
            raise TypeError(
                "Expected inputs to be dataclasses of the same type or to have identical fields"
                f" but got input 1 of type {type(data1)} and input 2 of type {type(data2)}"
            )
```

File: pytorch_lightning/utilities/apply_func.py [hunk 6]
```diff
@@ -222,6 +222,53 @@ def apply_to_collections(
...
+        assert len(dataclasses.fields(data1)) == len(dataclasses.fields(data2)) and map(
+            lambda f1, f2: isinstance(f1, type(f2)), zip(dataclasses.fields(data1), dataclasses.fields(data2))
+        ), "Dataclasses fields does not match"
```
Comment (awaelchli) [inline pytorch_lightning/utilities/apply_func.py:232]: same here :)

**Review by awaelchli — COMMENTED** [review]
looks great overall, just some minor comment

File: pytorch_lightning/utilities/apply_func.py (hunk 6 above)
Comment (armanal) [reply pytorch_lightning/utilities/apply_func.py]: Done ;)

File: tests/utilities/test_apply_func.py [hunk 7]
```diff
@@ -303,6 +308,36 @@ def fn(a, b):
...
+        feature=to_reduce_1,
+        label=torch.tensor([7.0, 8.0, 9.0]),
+    )
+
+    # different types
+    with pytest.raises(TypeError, match="Expected inputs to be dataclasses of the same type"):
+        apply_to_collections(to_reduce_1, [1, 2], (torch.Tensor, numbers.Number, np.ndarray), fn)
+
+    # unmatched fields
+    with pytest.raises(TypeError, match="Dataclasses fields does not match"):
```
Comment (awaelchli) [inline tests/utilities/test_apply_func.py:333]:
```suggestion
    with pytest.raises(TypeError, match="Dataclasses fields do not match"):
```

File: tests/utilities/test_apply_func.py [hunk 8]
```diff
@@ -303,6 +308,36 @@ def fn(a, b):
...
+        apply_to_collections(to_reduce_1, model_example, (torch.Tensor, numbers.Number, np.ndarray), fn)
+
+    classvar = WithClassVar(torch.arange(3))  # dataclass with same number but different type of fields
+    with pytest.raises(TypeError, match="Dataclasses fields does not match"):
```
Comment (awaelchli) [inline tests/utilities/test_apply_func.py:337]:
```suggestion
    with pytest.raises(TypeError, match="Dataclasses fields do not match"):
```

**Review by awaelchli — APPROVED** [review]
Great!

File: pytorch_lightning/utilities/apply_func.py [hunk 9]
```diff
@@ -222,6 +222,56 @@ def apply_to_collections(
...
+        if not (
+            len(dataclasses.fields(data1)) == len(dataclasses.fields(data2))
+            and map(lambda f1, f2: isinstance(f1, type(f2)), zip(dataclasses.fields(data1), dataclasses.fields(data2)))
+        ):
+            raise TypeError("Dataclasses fields do not match")
+        # make a deepcopy of the data,
+        # but do not deepcopy mapped fields since the computation would
+        # be wasted on values that likely get immediately overwritten
+        data = [data1, data2]
+        fields = [{}, {}]  # type: List[dict]
```
Comment (Borda) [inline pytorch_lightning/utilities/apply_func.py:240]: You can type it `fields: List[dict] = [{}, {}] `

File: pytorch_lightning/utilities/apply_func.py [hunk 10]
```diff
@@ -222,6 +222,56 @@ def apply_to_collections(
...
+        for i in range(len(data)):
+            for field in dataclasses.fields(data[i]):
+                field_value = getattr(data[i], field.name)
+                fields[i][field.name] = (field_value, field.init)
+                memo[i][id(field_value)] = field_value
+
+        result = deepcopy(data1, memo=memo[0])
+
+        # apply function to each field
+        for ((field_name, (field_value1, field_init1)), (_, (field_value2, field_init2))) in zip(
```
Comment (Borda) [inline pytorch_lightning/utilities/apply_func.py:251]: This for is confusing as it is designed to happen only once

File: tests/utilities/test_apply_func.py [hunk 11]
```diff
@@ -25,89 +25,94 @@
...
 
+@dataclasses.dataclass
+class Feature:
+    input_ids: torch.Tensor
+    segment_ids: np.ndarray
+
+    def __eq__(self, o: object) -> bool:
+        if not isinstance(o, Feature):
+            return NotImplemented
+        else:
```
Comment (Borda) [inline tests/utilities/test_apply_func.py:36]:
```suggestion
```

File: tests/utilities/test_apply_func.py [hunk 12]
```diff
@@ -25,89 +25,94 @@
...
+            return torch.equal(self.input_ids, o.input_ids) and np.equal(self.segment_ids, o.segment_ids).all()
```
Comment (Borda) [inline tests/utilities/test_apply_func.py:37]:
```suggestion
        return torch.equal(self.input_ids, o.input_ids) and np.equal(self.segment_ids, o.segment_ids).all()
```

Comment (Borda) [reply tests/utilities/test_apply_func.py:37]: The same bellow

File: pytorch_lightning/utilities/apply_func.py [hunk 13]
```diff
@@ -222,6 +222,56 @@ def apply_to_collections(
...
+        data = [data1, data2]
+        fields = [{}, {}]  # type: List[dict]
+        memo = [{}, {}]  # type: List[dict]
+        for i in range(len(data)):
+            for field in dataclasses.fields(data[i]):
+                field_value = getattr(data[i], field.name)
+                fields[i][field.name] = (field_value, field.init)
+                memo[i][id(field_value)] = field_value
+
+        result = deepcopy(data1, memo=memo[0])
```
Comment (rohitgr7) [inline pytorch_lightning/utilities/apply_func.py:248]: if `memo[1]` is not required, should we avoid computing for it?

File: pytorch_lightning/utilities/apply_func.py [hunk 14]
```diff
@@ -222,6 +222,56 @@ def apply_to_collections(
...
+
+        # apply function to each field
+        for ((field_name, (field_value1, field_init1)), (_, (field_value2, field_init2))) in zip(
+            fields[0].items(), fields[1].items()
+        ):
+            if field_init1 and field_init2:
```
Comment (rohitgr7) [inline pytorch_lightning/utilities/apply_func.py:254]: what if this condition fails, then `v` will not be defined, and the code below might fail.

File: pytorch_lightning/utilities/apply_func.py (hunk 10 above)
Comment (armanal) [reply pytorch_lightning/utilities/apply_func.py:251]: sorry, but I didn't get this one. what should I do about it?
it is similar to the one in `apply_to_collection`.

https://github.com/PyTorchLightning/pytorch-lightning/blob/2060fdec89d7f4585f122d440c252ead9340c519/pytorch_lightning/utilities/apply_func.py#L139-L159

File: pytorch_lightning/utilities/apply_func.py (hunk 14 above)
Comment (armanal) [reply pytorch_lightning/utilities/apply_func.py:254]:
```suggestion
            v = None
            if field_init1 and field_init2:
```

I also applied the same change here in `apply_to_collection`

https://github.com/PyTorchLightning/pytorch-lightning/blob/2060fdec89d7f4585f122d440c252ead9340c519/pytorch_lightning/utilities/apply_func.py#L140-L143

**Review by rohitgr7 — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

**Pull request merged**